In [1]:
import pandas as pd
import json

# Read the CSV file
df = pd.read_csv(
    "keywords.csv", 
    delimiter=',',  # Specify the delimiter (e.g., `;` if not `,`)
    quotechar='"',  # Specify the quote character
    encoding='utf-8',  # Set encoding to UTF-8
    skipinitialspace=True  # Skip spaces after delimiters
)



In [2]:
df

id                                           keywords
0         862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1        8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...
2       15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...
3       31357  [{'id': 818, 'name': 'based on novel'}, {'id':...
4       11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...
...       ...                                                ...
46414  439050             [{'id': 10703, 'name': 'tragic love'}]
46415  111109  [{'id': 2679, 'name': 'artist'}, {'id': 14531,...
46416   67758                                                 []
46417  227506                                                 []
46418  461257                                                 []

[46419 rows x 2 columns]

# 1.Data Exploration and Preprocessing

## crew, cast

this code processess the json columns (crew and cast) but due to their size and hardware limitation, I am not able to process them. <br>
we can use the same code for other json columns.

In [3]:
# import pandas as pd
# import ast

# # Load the CSV file
# file_path = "credits.csv"  # Replace with your CSV file path
# df = pd.read_csv(file_path)

# # Specify the column containing the nested JSON-like data
# column_name = ['crew' , 'cast']  # Replace with your actual column name

# for i in column_name:
#     # Convert the string representation of the JSON-like data into Python objects
#     df[i] = df[i].apply(ast.literal_eval)
    
#     # Flatten the JSON-like column
#     # Explode the list of dictionaries into separate rows
#     df = df.explode(i, ignore_index=True)
    
#     # Normalize the JSON objects and merge them with the original DataFrame
#     normalized_data = pd.json_normalize(df[i])
#     df = pd.concat([df.drop(columns=[i]), normalized_data], axis=1)

# # Save or display the resulting DataFrame
# df.to_csv("updated_file.csv", index=False)
# print(df.head())


## keywords.csv

In [4]:
import pandas as pd
import ast

# Load the CSV file
file_path = "keywords.csv"  # Replace with your CSV file path
df = pd.read_csv(file_path)

# Specify the column containing the nested JSON-like data
column_name = ['keywords']  # Replace with your actual column name

for i in column_name:
    # Convert the string representation of the JSON-like data into Python objects
    df[i] = df[i].apply(ast.literal_eval)
    
    # Flatten the JSON-like column
    # Explode the list of dictionaries into separate rows
    df = df.explode(i, ignore_index=True)
    
    # Normalize the JSON objects and merge them with the original DataFrame
    normalized_data = pd.json_normalize(df[i])
    normalized_data.rename(columns={'id':'word_id', 'name' : 'word'}, inplace =True )
    df = pd.concat([df.drop(columns=[i]), normalized_data], axis=1)

# Save or display the resulting DataFrame
df.dropna(inplace = True)
df.to_csv("updated_keywords.csv", index=False)
print(df.head())


    id  word_id        word
0  862    931.0    jealousy
1  862   4290.0         toy
2  862   5202.0         boy
3  862   6054.0  friendship
4  862   9713.0     friends


In [5]:
k_words = pd.read_csv('updated_keywords.csv')

In [6]:
k_words

id   word_id          word
0          862     931.0      jealousy
1          862    4290.0           toy
2          862    5202.0           boy
3          862    6054.0    friendship
4          862    9713.0       friends
...        ...       ...           ...
158675  289923   11800.0  mockumentary
158676  439050   10703.0   tragic love
158677  111109    2679.0        artist
158678  111109   14531.0          play
158679  111109  215397.0         pinoy

[158680 rows x 3 columns]

## rating_S.csv

In [7]:
rating_s = pd.read_csv('ratings_small.csv')
rating_s.drop(columns = ['timestamp'], inplace = True)
rating_s.head()


userId  movieId  rating
0       1       31     2.5
1       1     1029     3.0
2       1     1061     3.0
3       1     1129     2.0
4       1     1172     4.0

## movies_metadata.csv

In [8]:
movies_md = pd.read_csv('movies_metadata.csv', low_memory = False)
movies_md.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [9]:
movies_md.shape[0],movies_md.isnull().sum()

(45466,
 adult                        0
 belongs_to_collection    40972
 budget                       0
 genres                       0
 homepage                 37684
 id                           0
 imdb_id                     17
 original_language           11
 original_title               0
 overview                   954
 popularity                   5
 poster_path                386
 production_companies         3
 production_countries         3
 release_date                87
 revenue                      6
 runtime                    263
 spoken_languages             6
 status                      87
 tagline                  25054
 title                        6
 video                        6
 vote_average                 6
 vote_count                   6
 dtype: int64)

most of the movies do not have a homepage or do not belonge to a collection so this data is reduntant. <br>
information like imdb_id, overview, poster_path, tagline, video, title, status, original_title, adult does not help us neither.

In [10]:
movies_md.drop([19730,29503,35587],inplace = True) # input error
movies_md.drop(columns = ['belongs_to_collection','homepage','imdb_id', 'overview', 'poster_path', 'tagline', 'video', 'title', 'status', 'original_title', 'adult'], inplace = True)

In [11]:
movies_md.isnull().sum()

budget                    0
genres                    0
id                        0
original_language        11
popularity                3
production_companies      3
production_countries      3
release_date             87
revenue                   3
runtime                 260
spoken_languages          3
vote_average              3
vote_count                3
dtype: int64

In [12]:
movies_md.head()

budget                                             genres     id  \
0  30000000  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...    862   
1  65000000  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   8844   
2         0  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  15602   
3  16000000  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  31357   
4         0                     [{'id': 35, 'name': 'Comedy'}]  11862   

  original_language popularity  \
0                en  21.946943   
1                en  17.015539   
2                en    11.7129   
3                en   3.859495   
4                en   8.387519   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3  [{'name': 'Twentieth Century Fox Film Corporat...   
4  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   
2  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
3  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
4  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-02-10   

       revenue  runtime                                   spoken_languages  \
0  373554033.0     81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0    101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0    127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0    106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

   vote_average  vote_count  
0           7.7      5415.0  
1           6.9      2413.0  
2           6.5        92.0  
3           6.1        34.0  
4           5.7       173.0

replace genres with a list of genres ids

In [13]:
movies_md['genres'] = movies_md['genres'].apply(lambda x: [item['id'] for item in ast.literal_eval(x)])

In [14]:
movies_md.head()

budget           genres     id original_language popularity  \
0  30000000  [16, 35, 10751]    862                en  21.946943   
1  65000000  [12, 14, 10751]   8844                en  17.015539   
2         0      [10749, 35]  15602                en    11.7129   
3  16000000  [35, 18, 10749]  31357                en   3.859495   
4         0             [35]  11862                en   8.387519   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3  [{'name': 'Twentieth Century Fox Film Corporat...   
4  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   
2  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
3  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
4  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-02-10   

       revenue  runtime                                   spoken_languages  \
0  373554033.0     81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0    101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0    127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0    106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

   vote_average  vote_count  
0           7.7      5415.0  
1           6.9      2413.0  
2           6.5        92.0  
3           6.1        34.0  
4           5.7       173.0

do the same for production_companies, production countries and spoken languages. some of the data is missing, we replace the missing data with [].

In [15]:
from functools import partial

def extract_ids(missing, key, value):
    if pd.isna(value) or value == '' or len(value) == 0:
        return [missing]
    try:
        # Safely parse the string into a list of dictionaries and extract the ids
        parsed_value = ast.literal_eval(value)
        return [item[key] for item in parsed_value]
    except:
        # In case of error (invalid format), return [-1]
        return [missing]


f_companey = partial(extract_ids,-1, 'id')
movies_md['production_companies'] = movies_md['production_companies'].apply(f_companey)

f_country = partial(extract_ids,'US','iso_3166_1')
movies_md['production_countries'] = movies_md['production_countries'].apply(f_country)

f_lang = partial(extract_ids,'en','iso_639_1')
movies_md['spoken_languages'] = movies_md['spoken_languages'].apply(f_lang)


In [16]:
movies_md.head()

budget           genres     id original_language popularity  \
0  30000000  [16, 35, 10751]    862                en  21.946943   
1  65000000  [12, 14, 10751]   8844                en  17.015539   
2         0      [10749, 35]  15602                en    11.7129   
3  16000000  [35, 18, 10749]  31357                en   3.859495   
4         0             [35]  11862                en   8.387519   

  production_companies production_countries release_date      revenue  \
0                  [3]                 [US]   1995-10-30  373554033.0   
1   [559, 2550, 10201]                 [US]   1995-12-15  262797249.0   
2        [6194, 19464]                 [US]   1995-12-22          0.0   
3                [306]                 [US]   1995-12-22   81452156.0   
4         [5842, 9195]                 [US]   1995-02-10   76578911.0   

   runtime spoken_languages  vote_average  vote_count  
0     81.0             [en]           7.7      5415.0  
1    104.0         [en, fr]           6.9      2413.0  
2    101.0             [en]           6.5        92.0  
3    127.0             [en]           6.1        34.0  
4    106.0             [en]           5.7       173.0

# Recomendation

## Popularity-based filtering

We use [IMDB's Weighted Rating formula](https://en.wikipedia.org/wiki/IMDb#Rankings). <br>
$$ W = \frac{R \cdot v + C \cdot m}{v + m} $$
where: <br>
* w is the weighted raiting. <br>
* R is the mean rating for the movie, from 1 to 10.
* v is the number of votes for the movie.
* m is the minimum votes required to be listed (We chose 99% mark which is 2184)
* C is the mean vote across the whole report (110).

In our case, the lowest score possible is 110.

In [17]:
pop_df = movies_md.loc[:,['id','vote_average','vote_count']]
pop_df.head()

id  vote_average  vote_count
0    862           7.7      5415.0
1   8844           6.9      2413.0
2  15602           6.5        92.0
3  31357           6.1        34.0
4  11862           5.7       173.0

In [18]:
pop_df.isna().sum()

id              0
vote_average    3
vote_count      3
dtype: int64

In [19]:
pop_df.dropna(inplace = True)
pop_df.describe(),pop_df.vote_count.quantile(0.99)

(       vote_average    vote_count
 count  45460.000000  45460.000000
 mean       5.618207    109.897338
 std        1.924216    491.310374
 min        0.000000      0.000000
 25%        5.000000      3.000000
 50%        6.000000     10.000000
 75%        6.800000     34.000000
 max       10.000000  14075.000000,
 2183.8199999999924)

In [20]:
def w_rating(r , v ):
    c = 110
    m = 2184
    return ((r*v) + (c*m)) / (v+m)

In [21]:
w_rating(0,0)

110.0

In [22]:
pop_df['w_rating'] = w_rating(pop_df['vote_average'], pop_df['vote_count'])

In [23]:
pop_df

id  vote_average  vote_count    w_rating
0         862           7.7      5415.0   37.101658
1        8844           6.9      2413.0   55.882032
2       15602           6.5        92.0  105.816344
3       31357           6.1        34.0  108.407304
4       11862           5.7       173.0  102.344548
...       ...           ...         ...         ...
45461  439050           4.0         1.0  109.951487
45462  111109           9.0         3.0  109.861454
45463   67758           3.8         6.0  109.709041
45464  227506           0.0         0.0  110.000000
45465  461257           0.0         0.0  110.000000

[45460 rows x 4 columns]

In [24]:
sorted_df = pop_df.sort_values(by='w_rating', ascending=True)
sorted_df

id  vote_average  vote_count    w_rating
15480   27205           8.1     14075.0   21.787779
14551   19995           7.2     12114.0   22.902560
17818   24428           7.4     12000.0   23.197970
12481     155           8.3     12269.0   23.667937
26564  293660           7.4     11444.0   23.842501
...       ...           ...         ...         ...
34133   79506           0.0         0.0  110.000000
34115   77860           0.0         0.0  110.000000
34108  157724           0.0         0.0  110.000000
34105  156988           0.0         0.0  110.000000
45465  461257           0.0         0.0  110.000000

[45460 rows x 4 columns]

so the movie with 27205 id is the best movie.

In [98]:
raw_md = pd.read_csv('movies_metadata.csv', low_memory = False)
raw_md.id = raw_md.id.astype(str)
merged_df = sorted_df.merge(raw_md[['id', 'title']], on='id', how='left')
merged_df

id  vote_average  vote_count    w_rating  \
0       27205           8.1     14075.0   21.787779   
1       19995           7.2     12114.0   22.902560   
2       24428           7.4     12000.0   23.197970   
3         155           8.3     12269.0   23.667937   
4      293660           7.4     11444.0   23.842501   
...       ...           ...         ...         ...   
45517   79506           0.0         0.0  110.000000   
45518   77860           0.0         0.0  110.000000   
45519  157724           0.0         0.0  110.000000   
45520  156988           0.0         0.0  110.000000   
45521  461257           0.0         0.0  110.000000   

                                      title  
0                                 Inception  
1                                    Avatar  
2                              The Avengers  
3                           The Dark Knight  
4                                  Deadpool  
...                                     ...  
45517  Getting Away with It the Italian Way  
45518                        The Comedy Man  
45519                       The Crazy Bunch  
45520               The Return of Halleluja  
45521                              Queerama  

[45522 rows x 5 columns]

Its title is inception

## Content-based filtering

In [99]:
k_words

id   word_id          word
0          862     931.0      jealousy
1          862    4290.0           toy
2          862    5202.0           boy
3          862    6054.0    friendship
4          862    9713.0       friends
...        ...       ...           ...
158675  289923   11800.0  mockumentary
158676  439050   10703.0   tragic love
158677  111109    2679.0        artist
158678  111109   14531.0          play
158679  111109  215397.0         pinoy

[158680 rows x 3 columns]

for faster calculation we only use first 50000 movies

In [100]:
#binary_matrix = k_merged.pivot_table(index='id', columns='word', aggfunc='size', fill_value=0)
binary_matrix = k_words[:100000].pivot_table(index='id', columns='word', aggfunc='size', fill_value=0)
binary_matrix

word    1500s  15th birthday  15th century  16th century  17th century  \
id                                                                       
2           0              0             0             0             0   
3           0              0             0             0             0   
5           0              0             0             0             0   
6           0              0             0             0             0   
11          0              0             0             0             0   
...       ...            ...           ...           ...           ...   
415072      0              0             0             0             0   
429876      0              0             0             0             0   
438108      0              0             0             0             0   
440145      0              0             0             0             0   
458961      0              0             0             0             0   

word    18th century  1910s  1920s  1930s  1940s  ...   wichita kansas  \
id                                                ...                    
2                  0      0      0      0      0  ...                0   
3                  0      0      0      0      0  ...                0   
5                  0      0      0      0      0  ...                0   
6                  0      0      0      0      0  ...                0   
11                 0      0      0      0      0  ...                0   
...              ...    ...    ...    ...    ...  ...              ...   
415072             0      0      0      0      0  ...                0   
429876             0      0      0      0      0  ...                0   
438108             0      0      0      0      0  ...                0   
440145             0      0      0      0      0  ...                0   
458961             0      0      0      0      0  ...                0   

word    Åsa-nisse  Øverste hylde  Γη  איים אבודים  ישראלי  卧底肥妈  绝地奶霸  超级妈妈  \
id                                                                            
2               0              0   0            0       0     0     0     0   
3               0              0   0            0       0     0     0     0   
5               0              0   0            0       0     0     0     0   
6               0              0   0            0       0     0     0     0   
11              0              0   0            0       0     0     0     0   
...           ...            ...  ..          ...     ...   ...   ...   ...   
415072          0              0   0            0       0     0     0     0   
429876          0              0   0            0       0     0     0     0   
438108          0              0   0            0       0     0     0     0   
440145          0              0   0            0       0     0     0     0   
458961          0              0   0            0       0     0     0     0   

word    형사 duelist  
id                  
2                0  
3                0  
5                0  
6                0  
11               0  
...            ...  
415072           0  
429876           0  
438108           0  
440145           0  
458961           0  

[16654 rows x 16087 columns]

In [101]:
from sklearn.metrics.pairwise import cosine_similarity

# Compute cosine similarity
cosine_sim = cosine_similarity(binary_matrix)

# Convert to DataFrame for better readability
cosine_sim_df = pd.DataFrame(cosine_sim, index=binary_matrix.index, columns=binary_matrix.index)
cosine_sim_df


id        2         3         5       6       11      12      13      14      \
id                                                                             
2       1.000000  0.377964  0.000000     0.0     0.0     0.0     0.0     0.0   
3       0.377964  1.000000  0.000000     0.0     0.0     0.0     0.0     0.0   
5       0.000000  0.000000  1.000000     0.0     0.0     0.0     0.0     0.0   
6       0.000000  0.000000  0.000000     1.0     0.0     0.0     0.0     0.0   
11      0.000000  0.000000  0.000000     0.0     1.0     0.0     0.0     0.0   
...          ...       ...       ...     ...     ...     ...     ...     ...   
415072  0.000000  0.000000  0.000000     0.0     0.0     0.0     0.0     0.0   
429876  0.000000  0.000000  0.316228     0.0     0.0     0.0     0.0     0.0   
438108  0.000000  0.000000  0.000000     0.0     0.0     0.0     0.0     0.0   
440145  0.000000  0.000000  0.000000     0.0     0.0     0.0     0.0     0.0   
458961  0.000000  0.000000  0.000000     0.0     0.0     0.0     0.0     0.0   

id      15        16      ...  392067  405280    407992    408956    409900  \
id                        ...                                                 
2          0.0  0.070186  ...     0.0     0.0  0.169031  0.000000  0.000000   
3          0.0  0.000000  ...     0.0     0.0  0.223607  0.000000  0.000000   
5          0.0  0.000000  ...     0.0     0.0  0.141421  0.316228  0.316228   
6          0.0  0.000000  ...     0.0     0.0  0.000000  0.000000  0.000000   
11         0.0  0.000000  ...     0.0     0.0  0.000000  0.000000  0.000000   
...        ...       ...  ...     ...     ...       ...       ...       ...   
415072     0.0  0.000000  ...     0.0     0.0  0.000000  0.000000  0.000000   
429876     0.0  0.000000  ...     0.0     0.0  0.447214  1.000000  1.000000   
438108     0.0  0.000000  ...     0.0     0.0  0.000000  0.000000  0.000000   
440145     0.0  0.000000  ...     1.0     0.0  0.000000  0.000000  0.000000   
458961     0.0  0.000000  ...     0.0     0.0  0.000000  0.000000  0.000000   

id      415072    429876  438108  440145  458961  
id                                                
2          0.0  0.000000     0.0     0.0     0.0  
3          0.0  0.000000     0.0     0.0     0.0  
5          0.0  0.316228     0.0     0.0     0.0  
6          0.0  0.000000     0.0     0.0     0.0  
11         0.0  0.000000     0.0     0.0     0.0  
...        ...       ...     ...     ...     ...  
415072     1.0  0.000000     0.0     0.0     0.0  
429876     0.0  1.000000     0.0     0.0     0.0  
438108     0.0  0.000000     1.0     0.0     0.0  
440145     0.0  0.000000     0.0     1.0     0.0  
458961     0.0  0.000000     0.0     0.0     1.0  

[16654 rows x 16654 columns]

In [106]:
def recommend_movies(movie_id, df, top_n=3):
    # Get the similarity scores for the given movie
    sim_scores = df[movie_id].sort_values(ascending=False)

    # Exclude the input movie itself from the recommendations
    sim_scores = sim_scores[sim_scores.index != movie_id]

    # Get the top N most similar movies
    return sim_scores.head(top_n)

# Example usage: Recommend top 3 similar movies to movie 5

movie = 11
raw_md.loc[raw_md.id == str(movie)].title, raw_md.loc[raw_md.id.isin(recommend_movies(movie, cosine_sim_df, top_n=7).index.astype(str))].title

(256    Star Wars
 Name: title, dtype: object,
 1154                          The Empire Strikes Back
 1167                               Return of the Jedi
 5801                      The Man Who Saves The World
 10069    Star Wars: Episode III - Revenge of the Sith
 12296                     Battlestar Galactica: Razor
 12885                       Star Wars: The Clone Wars
 16175                              Little Red Flowers
 Name: title, dtype: object)

## Collaborative filtering

In [107]:
rating_s

userId  movieId  rating
0            1       31     2.5
1            1     1029     3.0
2            1     1061     3.0
3            1     1129     2.0
4            1     1172     4.0
...        ...      ...     ...
99999      671     6268     2.5
100000     671     6269     4.0
100001     671     6365     4.0
100002     671     6385     2.5
100003     671     6565     3.5

[100004 rows x 3 columns]

In [108]:
user_movie_matrix = rating_s.pivot_table(index='userId', columns='movieId', values='rating', fill_value=0)
user_movie_matrix

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
3           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
4           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
5           0.0     0.0     4.0     0.0     0.0     0.0     0.0     0.0   
...         ...     ...     ...     ...     ...     ...     ...     ...   
667         0.0     0.0     0.0     0.0     0.0     4.0     0.0     0.0   
668         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
669         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
670         4.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
671         5.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

movieId  9       10      ...  161084  161155  161594  161830  161918  161944  \
userId                   ...                                                   
1           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
2           0.0     4.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
3           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
4           0.0     4.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
5           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
...         ...     ...  ...     ...     ...     ...     ...     ...     ...   
667         0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
668         0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
669         0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
670         0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
671         0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   

movieId  162376  162542  162672  163949  
userId                                   
1           0.0     0.0     0.0     0.0  
2           0.0     0.0     0.0     0.0  
3           0.0     0.0     0.0     0.0  
4           0.0     0.0     0.0     0.0  
5           0.0     0.0     0.0     0.0  
...         ...     ...     ...     ...  
667         0.0     0.0     0.0     0.0  
668         0.0     0.0     0.0     0.0  
669         0.0     0.0     0.0     0.0  
670         0.0     0.0     0.0     0.0  
671         0.0     0.0     0.0     0.0  

[671 rows x 9066 columns]

### user based

In [109]:
from sklearn.metrics.pairwise import cosine_similarity

# Calculate similarity between users
user_similarity = cosine_similarity(user_movie_matrix)
user_similarity_df = pd.DataFrame(user_similarity, index=user_movie_matrix.index, columns=user_movie_matrix.index)
user_similarity_df


userId       1         2         3         4         5         6         7    \
userId                                                                         
1       1.000000  0.000000  0.000000  0.074482  0.016818  0.000000  0.083884   
2       0.000000  1.000000  0.124295  0.118821  0.103646  0.000000  0.212985   
3       0.000000  0.124295  1.000000  0.081640  0.151531  0.060691  0.154714   
4       0.074482  0.118821  0.081640  1.000000  0.130649  0.079648  0.319745   
5       0.016818  0.103646  0.151531  0.130649  1.000000  0.063796  0.095888   
...          ...       ...       ...       ...       ...       ...       ...   
667     0.000000  0.425462  0.124562  0.088735  0.058252  0.000000  0.232051   
668     0.000000  0.084646  0.124911  0.068483  0.042926  0.019563  0.058773   
669     0.062917  0.024140  0.080984  0.104309  0.038358  0.024583  0.073151   
670     0.000000  0.170595  0.136606  0.054512  0.062642  0.019465  0.096240   
671     0.017466  0.113175  0.170193  0.211609  0.225086  0.087705  0.268672   

userId       8         9         10   ...       662       663       664  \
userId                                ...                                 
1       0.000000  0.012843  0.000000  ...  0.000000  0.000000  0.014474   
2       0.113190  0.113333  0.043213  ...  0.477306  0.063202  0.077745   
3       0.249781  0.134475  0.114672  ...  0.161205  0.064198  0.176134   
4       0.191013  0.030417  0.137186  ...  0.114319  0.047228  0.136579   
5       0.165712  0.086616  0.032370  ...  0.191029  0.021142  0.146173   
...          ...       ...       ...  ...       ...       ...       ...   
667     0.069005  0.066412  0.032653  ...  0.342283  0.050754  0.074080   
668     0.112366  0.194493  0.098561  ...  0.074089  0.059010  0.093021   
669     0.055143  0.029291  0.060549  ...  0.015960  0.025953  0.077927   
670     0.247687  0.384429  0.158650  ...  0.183662  0.122126  0.123407   
671     0.406414  0.168497  0.189703  ...  0.096704  0.070749  0.219728   

userId       665       666       667       668       669       670       671  
userId                                                                        
1       0.043719  0.000000  0.000000  0.000000  0.062917  0.000000  0.017466  
2       0.164162  0.466281  0.425462  0.084646  0.024140  0.170595  0.113175  
3       0.158357  0.177098  0.124562  0.124911  0.080984  0.136606  0.170193  
4       0.254030  0.121905  0.088735  0.068483  0.104309  0.054512  0.211609  
5       0.224245  0.139721  0.058252  0.042926  0.038358  0.062642  0.225086  
...          ...       ...       ...       ...       ...       ...       ...  
667     0.124214  0.351207  1.000000  0.091597  0.018416  0.153111  0.127995  
668     0.082525  0.114487  0.091597  1.000000  0.000000  0.178017  0.135387  
669     0.101707  0.028773  0.018416  0.000000  1.000000  0.042609  0.085202  
670     0.143380  0.159479  0.153111  0.178017  0.042609  1.000000  0.228677  
671     0.239318  0.127204  0.127995  0.135387  0.085202  0.228677  1.000000  

[671 rows x 671 columns]

In [111]:
recommend_movies(1, user_similarity_df, top_n=10)

userId
325    0.371852
634    0.194093
341    0.162819
310    0.157524
207    0.152746
35     0.130585
195    0.122647
485    0.114021
130    0.112817
229    0.112575
Name: 1, dtype: float64

### movie based

In [112]:
# Calculate similarity between movies
movie_similarity = cosine_similarity(user_movie_matrix.T)
movie_similarity_df = pd.DataFrame(movie_similarity, index=user_movie_matrix.columns, columns=user_movie_matrix.columns)
movie_similarity_df


movieId    1         2         3         4         5         6         7       \
movieId                                                                         
1        1.000000  0.394511  0.306516  0.133614  0.245102  0.377086  0.278629   
2        0.394511  1.000000  0.217492  0.164651  0.278476  0.222003  0.207299   
3        0.306516  0.217492  1.000000  0.177012  0.370732  0.247499  0.435648   
4        0.133614  0.164651  0.177012  1.000000  0.179556  0.072518  0.184626   
5        0.245102  0.278476  0.370732  0.179556  1.000000  0.272645  0.388476   
...           ...       ...       ...       ...       ...       ...       ...   
161944   0.079755  0.137594  0.000000  0.000000  0.000000  0.000000  0.000000   
162376   0.079755  0.068797  0.000000  0.000000  0.000000  0.111103  0.000000   
162542   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
162672   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
163949   0.055829  0.000000  0.000000  0.000000  0.000000  0.061724  0.079399   

movieId    8         9         10      ...    161084    161155    161594  \
movieId                                ...                                 
1        0.063031  0.117499  0.310689  ...  0.055829  0.031902  0.079755   
2        0.223524  0.113669  0.418124  ...  0.000000  0.055038  0.068797   
3        0.127574  0.306717  0.191255  ...  0.000000  0.000000  0.000000   
4        0.501513  0.254630  0.111447  ...  0.000000  0.000000  0.000000   
5        0.194113  0.367941  0.246846  ...  0.000000  0.176845  0.000000   
...           ...       ...       ...  ...       ...       ...       ...   
161944   0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
162376   0.000000  0.000000  0.076835  ...  0.000000  0.000000  1.000000   
162542   0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
162672   0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
163949   0.000000  0.000000  0.000000  ...  1.000000  0.000000  0.000000   

movieId    161830    161918    161944    162376  162542  162672    163949  
movieId                                                                    
1        0.079755  0.079755  0.079755  0.079755     0.0     0.0  0.055829  
2        0.082557  0.082557  0.137594  0.068797     0.0     0.0  0.000000  
3        0.116226  0.116226  0.000000  0.000000     0.0     0.0  0.000000  
4        0.000000  0.000000  0.000000  0.000000     0.0     0.0  0.000000  
5        0.117897  0.117897  0.000000  0.000000     0.0     0.0  0.000000  
...           ...       ...       ...       ...     ...     ...       ...  
161944   0.000000  0.000000  1.000000  0.000000     0.0     0.0  0.000000  
162376   0.000000  0.000000  0.000000  1.000000     0.0     0.0  0.000000  
162542   0.000000  0.000000  0.000000  0.000000     1.0     1.0  0.000000  
162672   0.000000  0.000000  0.000000  0.000000     1.0     1.0  0.000000  
163949   0.000000  0.000000  0.000000  0.000000     0.0     0.0  1.000000  

[9066 rows x 9066 columns]

In [113]:
recommend_movies(1, movie_similarity_df, top_n=5)

movieId
3114    0.594710
260     0.576188
356     0.564534
780     0.562946
1265    0.548023
Name: 1, dtype: float64

In [114]:
movie = 11
raw_md.loc[raw_md.id == str(movie)].title, raw_md.loc[raw_md.id.isin(recommend_movies(movie, movie_similarity_df, top_n=10).index.astype(str))].title

(256    Star Wars
 Name: title, dtype: object,
 1042     Willy Wonka & the Chocolate Factory
 1176                                  Psycho
 1234                          Night on Earth
 1299               A Nightmare on Elm Street
 1639                                 Titanic
 1852                                Rain Man
 7275                              Young Adam
 12284            Aliens vs Predator: Requiem
 Name: title, dtype: object)

# SVD

In [84]:
from sklearn.decomposition import TruncatedSVD

# Apply SVD
svd = TruncatedSVD(n_components=2)
decomposed_matrix = svd.fit_transform(user_movie_matrix.T)

# Compute similarity between users in the reduced space
movie_similarity_svd = cosine_similarity(decomposed_matrix)
movie_similarity_svd_df = pd.DataFrame(movie_similarity_svd, index=user_movie_matrix.columns, columns=user_movie_matrix.columns)
movie_similarity_svd_df


movieId    1         2         3         4         5         6         7       \
movieId                                                                         
1        1.000000  0.999921  0.916736  0.773521  0.945381  0.972229  0.726885   
2        0.999921  1.000000  0.911650  0.765506  0.941216  0.969215  0.718210   
3        0.916736  0.911650  1.000000  0.962302  0.996887  0.984772  0.940718   
4        0.773521  0.765506  0.962302  1.000000  0.937860  0.900362  0.997509   
5        0.945381  0.941216  0.996887  0.937860  1.000000  0.995414  0.911044   
...           ...       ...       ...       ...       ...       ...       ...   
161944   0.694630  0.703603  0.349409  0.081396  0.422200  0.506983  0.010884   
162376   0.872937  0.878990  0.605366  0.366060  0.666241  0.734525  0.299501   
162542   0.627311  0.637035  0.263965 -0.008323  0.339196  0.427632 -0.078843   
162672   0.627311  0.637035  0.263965 -0.008323  0.339196  0.427632 -0.078843   
163949   0.495708  0.484770  0.801389  0.933864  0.751731  0.685197  0.956765   

movieId    8         9         10      ...    161084    161155    161594  \
movieId                                ...                                 
1        0.651249  0.750265  0.999627  ...  0.495708  0.994467  0.872937   
2        0.641674  0.741909  0.999891  ...  0.484770  0.993071  0.878990   
3        0.900184  0.951915  0.905482  ...  0.801389  0.953630  0.605366   
4        0.984701  0.999355  0.755920  ...  0.933864  0.835817  0.366060   
5        0.863042  0.924795  0.936125  ...  0.751731  0.974393  0.666241   
...           ...       ...       ...  ...       ...       ...       ...   
161944  -0.093526  0.045556  0.714020  ... -0.280430  0.615219  0.957299   
162376   0.198300  0.332409  0.885936  ...  0.009044  0.816862  1.000000   
162542  -0.182445 -0.044223  0.648349  ... -0.365389  0.542033  0.927512   
162672  -0.182445 -0.044223  0.648349  ... -0.365389  0.542033  0.927512   
163949   0.981895  0.946103  0.471800  ...  1.000000  0.584198  0.009044   

movieId    161830    161918    161944    162376    162542    162672    163949  
movieId                                                                        
1        0.976754  0.976754  0.694630  0.872937  0.627311  0.627311  0.495708  
2        0.979367  0.979367  0.703603  0.878990  0.637035  0.637035  0.484770  
3        0.809789  0.809789  0.349409  0.605366  0.263965  0.263965  0.801389  
4        0.619682  0.619682  0.081396  0.366060 -0.008323 -0.008323  0.933864  
5        0.853530  0.853530  0.422200  0.666241  0.339196  0.339196  0.751731  
...           ...       ...       ...       ...       ...       ...       ...  
161944   0.832688  0.832688  1.000000  0.957299  0.995970  0.995970 -0.280430  
162376   0.957218  0.957218  0.957299  1.000000  0.927512  0.927512  0.009044  
162542   0.779668  0.779668  0.995970  0.927512  1.000000  1.000000 -0.365389  
162672   0.779668  0.779668  0.995970  0.927512  1.000000  1.000000 -0.365389  
163949   0.298013  0.298013 -0.280430  0.009044 -0.365389 -0.365389  1.000000  

[9066 rows x 9066 columns]

In [135]:
movie = 11
raw_md.loc[raw_md.id == str(movie)].title, raw_md.loc[raw_md.id.isin(recommend_movies(movie, movie_similarity_svd_df, top_n=20).index.astype(str))].title

(256    Star Wars
 Name: title, dtype: object,
 2160                              Rush Hour
 2419     Conquest of the Planet of the Apes
 2445                       Forces of Nature
 2869                                RoboCop
 5347            Austin Powers in Goldmember
 6522                   An Angel at My Table
 6938                                   Hero
 7877                                   Taxi
 11505          The Young Girls of Rochefort
 Name: title, dtype: object)

# hybrid

In [140]:
def hybrid_recommend(movie_id, content_df, collab_df, alpha=0.5, top_n=5):
    
    # Get content-based similarity scores for the movie
    content_scores = content_df[movie_id]
    
    # Get collaborative similarity scores for the movie
    collab_scores = collab_df[movie_id]
    
    # Combine scores using a weighted average
    hybrid_scores = (alpha * content_scores) + ((1 - alpha) * collab_scores)
    
    # Sort movies by hybrid scores
    recommendations = hybrid_scores.sort_values(ascending=False).drop(movie_id)
    
    # Return top N recommendations
    return recommendations.head(top_n)

# Example: Recommend movies similar to movie 101
hybrid_recommend(101, cosine_sim_df, movie_similarity_df, alpha=0.5, top_n=10)

movie = 11
raw_md.loc[raw_md.id == str(movie)].title, raw_md.loc[raw_md.id.isin(hybrid_recommend(movie, cosine_sim_df, movie_similarity_df, alpha=0.5, top_n=10).index.astype(str))].title

(256    Star Wars
 Name: title, dtype: object,
 1042              Willy Wonka & the Chocolate Factory
 1154                          The Empire Strikes Back
 1167                               Return of the Jedi
 1176                                           Psycho
 1234                                   Night on Earth
 1299                        A Nightmare on Elm Street
 1639                                          Titanic
 1852                                         Rain Man
 5244     Star Wars: Episode II - Attack of the Clones
 12284                     Aliens vs Predator: Requiem
 Name: title, dtype: object)